save-frontend/src/main/kotlin/com/saveourtool/save/frontend/components/views/url/ParseUrInTab.kt

```diff
@@ -0,0 +1,69 @@
+package com.saveourtool.save.frontend.components.views.url
+
+import com.saveourtool.save.domain.Role
+import com.saveourtool.save.entities.benchmarks.TabMenubar
+import com.saveourtool.save.frontend.components.views.AbstractView
+import com.saveourtool.save.utils.URL_PATH_DELIMITER
+import com.saveourtool.save.validation.FrontendRoutes
+import kotlinx.browser.window
+import react.State
+
+
+/**
+ * The function of analyzing the URL of a tabbed page goes to the tab that was entered in the url, according to the role
+ *
+ * @param menu
+ * @param role
+ * @param flag
+ */
+fun <T, S> AbstractView<*, S>.urlAnalysis(menu: TabMenubar<T>, role: Role, flag: Boolean?)
+        where S : State, S : HasSelectedMenu<T> {
+    val href = window.location.href
+    val tab = if (href.contains(menu.regexForUrlClassification)) {
+        href.substringAfterLast(URL_PATH_DELIMITER).run { menu.findEnumElements(this) }
+    } else {
+        menu.defaultTab
+    }
+    if (state.selectedMenu != tab) {
+        if (menu.isNotAvailableWithThisRole(role, tab, flag)) {
+            changeUrl(menu.defaultTab, menu)
+            window.alert("Your role is not suitable for opening this page")
+            window.location.reload()
+            setState { selectedMenu = menu.defaultTab }
+        } else {
+            changeUrl(tab, menu)
+            setState { selectedMenu = tab }
+        }
+    }
+}
+
+/**
+ * Creates unique url address for page tabs
+ *
+ * @param selectedMenu
+ * @param menubar
+ */
+fun <T, S> changeUrl(selectedMenu: T?, menubar: TabMenubar<T>)
+        where S : HasSelectedMenu<T> {
+    selectedMenu?.let {
+        window.location.href = if (selectedMenu == menubar.defaultTab) {
+            menubar.paths.first
```

Comment:
@Arrgentum using `.first` and `.second` is usually only suitable when declaration of the pair is right near its usage. Otherwise you might've forgotten what corresponds to what and also there is no control of how these fields are assigned (you are calling `setPath` in the middle of the code, for example). A small data class with named fields is more suitable here